Add a guide on input validation (#44287)
mdn/content: files/en-us/web/security/defenses/input_validation/index.md

---
title: Input validation
slug: Web/Security/Defenses/Input_validation
page-type: guide
sidebar: security
---

Input validation is the practice of checking that any input that your website accepts is what you expect it to be.

> For a website to provide almost any kind of interactivity or customization it will need to accept some input, typically from users via a web browser, but also sometimes from other applications.

Reviewer: FWIW Was not wrong, but this felt better to me. 

```suggestion
For a website to provide almost any kind of interactivity or customization it will need to accept some input, typically from users via a web browser, and sometimes also from other applications.
```

> Users typically provide input using {{htmlelement("input")}} elements in a {{htmlelement("form")}} element on the site's front end, and the input is typically sent to the server as the body of a {{httpmethod("POST")}} request, or as URL parameters appended to a {{httpmethod("GET")}} request. However, input might also arrive on the server via other means, such as cookie values or additional HTTP headers.

Reviewer: Nitty, but I'm already sick of input

```suggestion
Users typically enter information into {{htmlelement("input")}} elements in a {{htmlelement("form")}} element on the site's front end, and the data is typically sent to the server as the body of a {{httpmethod("POST")}} request, or as URL parameters appended to a {{httpmethod("GET")}} request. However, input might also arrive on the server via other means, such as cookie values or additional HTTP headers.
```

> If the input provided by the user doesn't have the form or content that the server expects — for example, if they enter an invalid email address — this can cause the site to malfunction. Catching problems like this as early as possible improves the user's experience.

Reviewer: Feels quite server centric. But I do think it is important to highlight both cases of what is effectively catching user error vs malicious usage.

Author: "server"->"site" change is made in https://github.com/mdn/content/pull/44287/commits/8290e284452e959c585f646c5ef20e958e28cf66, which I agree is better, but otherwise I'm not sure what this comment is asking for.

Reviewer: What I was saying is that this is about server validation, but there is also validation on the client. Most of that client side validation is really about catching mistakes in user input, while the server testing exists primarily to catch malicious use.

Anyway, it's a bit of a ramble - wasn't really asking for anything except perhaps thinking "this is all about validation, but the client side stuff isn't being mentioned at all". Might be a very reasonable approach.

Author: But we do talk about client-side validation, a lot, notably in "when to validate" and the whole section on client-side validation. This preamble doesn't really talk about validation at all, on client or server, it's mostly concerned with where input comes from, how to gets into the server, and how bad input can cause problems.

Reviewer: @wbamberg Just to close this, you're right. 

I made this comment at the start of my last review before re-reading and forgot to come back at the end of the review and close it as "whatever my original reasons, no longer feel this way". Sorry.

However, aside from innocent errors by the user, supplying unexpected input enables a malicious actor to attempt various security attacks, including cross-site scripting (XSS), SQL injection, and command injection. In these situations, the attacker deliberately crafts some input that will enable an attack, and passes it to the application. The attacker might completely bypass the website's front-end, providing the malicious input directly in an HTTP request. Although input validation is not usually a complete solution to these security threats, it's an important first line of defense.

## Validation guidelines

### Implement validation as an allowlist

> Applications can often implement a given check as either an allowlist or a denylist.

Reviewer: So nitish, but this isn't a list. Perhaps we can define it as such.

```suggestion
Applications can often implement a given check by specifying it in terms of the values that are allowed (an "allowlist") or in terms of those that are denied (a "denylist").
```

> Applications should implement validation as an allowlist, rather than a denylist. That is, applications should allow only input that positively matches the expected value.

Reviewer: Pluralise?

```suggestion
Applications should implement validation as an allowlist, rather than a denylist. That is, applications should allow only input that positively matches expected values.
```

For example, suppose we wanted to check that a numeric input is between zero and 10. We can test that the input matches this range, and deny everything else:

```js
function checkRange(input) {
  if (input >= 0 && input <= 10) {
    return true;
  }
  return false;
}
```

Alternatively, we can test whether the input is outside the range, and allow everything else:

```js example-bad
function checkRange(input) {
  if (input < 0 || input > 10) {
    return false;
  }
  return true;
}
```

> This is much less safe, because it's easier for an attacker to craft some input that would evade the check, and fall through to the default "allow" condition.

Reviewer: The guidance is true, but FWIW I am not convinced by this example and final statement. 

Specifically I think you validate an allow list because it is much easier and safer to define what is good than to predict every single variation of what is bad. Further, what is bad can change over time, but if you know something is good, there is a chance that you can keep it forever. 

Hard to illustrate that with this example - there's no real way an attacker can evade this.

It just feels that this is stretching too hard to create an attack case. We're doing this mostly to help users not screw up their input, and we're using an allow list because psychologically humans seem better able to specify this is the allowed range than "this is not".

Not sure how to do it - the easy case for an attack  is HTML input validation but it is very different from a numeric range check - which has far more to do with what I think of when someone says "input validation"

Reviewer: PS I might be overthinking this. I just spent a while trying to craft  message that this draws quite a tenuous link between validation and protecting against malicious attack. But then I reread and saw both your intro  ...

> However, aside from innocent errors by the user, supplying unexpected input enables a malicious actor to attempt various security attacks, including cross-site scripting (XSS), SQL injection, and command injection. In these situations, the attacker deliberately crafts some input that will enable an attack, and passes it to the application. The attacker might completely bypass the website's front-end, providing the malicious input directly in an HTTP request. Although input validation is not usually a complete solution to these security threats, it's an important first line of defense.

And end comment ....

> Generic input validation as we've discussed it here is a useful first line of defense against attacks like this, but it is _not_ a complete defense against them, or even the primary defense. This is because it's extremely hard to protect against these attacks without knowing the specific context in which the input will be used.

That make this point.

Author: > FWIW I am not convinced by this example and final statement.

Yeah, I agree about this.

I kept the examples, as illustrations of "allowlist" and "denylist", but separated them from the guidance and the reasoning behind it. -> https://github.com/mdn/content/pull/44287/commits/8290e284452e959c585f646c5ef20e958e28cf66

> It's usually more reliable to implement a check as an allowlist, as this defaults to denying values which the author didn't consider. This is especially true when the invalid input is being deliberately crafted by an attacker: it's easier for an attacker to provide some input that would evade the check, and fall through to the default "allow" condition.

Reviewer: I _think_ it is that here.

```suggestion
It's usually more reliable to implement a check as an allowlist, as this defaults to denying values that the author didn't consider. This is especially true when the invalid input is being deliberately crafted by an attacker: it's easier for an attacker to provide some input that would evade the check, and fall through to the default "allow" condition.
```

### Syntactic and semantic validation

We can distinguish two types of input validation:

- _syntactic validation_, which checks that input is in the correct format. For example, if the application expects a number, it is receiving a number.
- _semantic validation_, which checks that the content of the input is within expected bounds. For example, a number that should be within a specific range, or a string value that should exactly match one of a set of values.

> Applications typically implement syntactic validation using type-checking features of the programming languages that they use.

Reviewer: ```suggestion
Applications typically implement syntactic validation using the type-checking features of their chosen programming language.
```

> To implement semantic validation, they can use various methods including range checking, checking a value against a set of allowed values, or, for more complex cases, regular expressions.

Reviewer: Edit
~Thinking out loud, it might be worth mentioning "how" here - that HTML has quite a lot of validation on inputs, such as the type, and max, min values.~ I see you did this two sections on.

EDIT ~But, that you also still need to validate on the server - Django etc provide libraries for that.~ - I see you did this in the next section

> Note that regular expressions can be hard to get right, and can make an application vulnerable to [denial of service attacks](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS) in some cases. For this reason, it's usually better to use well-regarded third party libraries here. One popular choice is [validator.js](https://github.com/validatorjs/validator.js).

Reviewer: Can already implies some cases. Maybe

```suggestion
Note that regular expressions can be hard to get right, and some expressions may make an application vulnerable to [denial of service attacks](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS). For this reason, it's usually better to use well-regarded third party validation libraries. One popular choice is [validator.js](https://github.com/validatorjs/validator.js).
```

See [Using HTML form validation](/en-US/docs/Web/HTML/Guides/Constraint_validation) for more information.

## Server-side validation

On the server side, applications should, if possible, use the validation functions provided by their framework of choice, such as Django's [validators](https://docs.djangoproject.com/en/6.0/ref/validators/).

It's especially important to pay attention to validation failures which could not have been made by a user interacting with the site's front end: for example, a {{htmlelement("select")}} element containing an option that was not provided in the form's HTML. Failures like this are strong indicators that an attacker is deliberately crafting invalid input.

When input is represented as JSON, you can use [JSON Schema](https://json-schema.org/) to define validity: this includes APIs that are specified using [OpenAPI](https://swagger.io/). If you use a database, you can also define a database schema and validate input against that.

## File uploads

If your site allows file uploads, you need to consider various additional threats. Attackers may:

- Upload malicious files that exploit bugs in the software that processes them.

> - Upload very large files, that create a denial of service attack.

Reviewer: Wording didn't seem quite right. Don't know if we need to say more here - i.e. just consuming resources that you're not supposed to can be a cost to the site. 

```suggestion
- Upload very large files as part of a denial of service attack.
```

- Confuse file-handling code into overwriting your own site's files.
- Upload files containing exploits such as XSS, and trick other users into downloading and executing them.

In this section we'll outline a few practices to help defend against these threats. For more details, see OWASP's [File Upload Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/File_Upload_Cheat_Sheet.html).

- Only allow authenticated users to upload files.

- Filenames are also user input, so they need to be validated. If possible, generate your own name for files you save. If you have to use the names provided by users, be highly restrictive in the characters that users may choose, and validate names against this.

- Determine which file types you need to support, and allow only these types, based on the file extension. Be especially wary of web-executable file types such as HTML or JavaScript. Since checking file extensions involves processing the filenames provided by users, be sure to validate filenames first.

- Restrict the size of files that can be uploaded.

- Allow users to report undesirable or illegal content, and have a process for removing it.

> - If possible, store files on a different host. If this is not possible, store files outside the website's root. This reduces the risk that malicious uploads could be served to other users in attacks such as XSS.

Reviewer: The "in this section" is usually used for larger sections. I might do it this way. Note text is the same except the intro and the OWASP line being pushed to the end.

```suggestion
The following best practices are commonly used to mitigate these threats:

- Only allow authenticated users to upload files.

- Filenames are also user input, so they need to be validated. If possible, generate your own name for files you save. If you have to use the names provided by users, be highly restrictive in the characters that users may choose, and validate names against this.

- Determine which file types you need to support, and allow only these types, based on the file extension. Be especially wary of web-executable file types such as HTML or JavaScript. Since checking file extensions involves processing the filenames provided by users, be sure to validate filenames first.

- Restrict the size of files that can be uploaded.

- Allow users to report undesirable or illegal content, and have a process for removing it.

- If possible, store files on a different host. If this is not possible, store files outside the website's root. This reduces the risk that malicious uploads could be served to other users in attacks such as XSS.

For more details, see OWASP's [File Upload Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/File_Upload_Cheat_Sheet.html).
```
